File: HashmapAndSet/missing-number.js
/*Given an array nums containing n distinct numbers in the range [0, n], 
return the only number in the range that is missing from the array.*/

//sorting solution to start
const missingNumber = (nums) => {
    const sortedArr = nums.sort((a, b) => a - b);

    if (sortedArr[0] !== 0) return 0;
    if (sortedArr[sortedArr.length - 1] !== sortedArr.length) return sortedArr.length;

    for (let i = 1; i < sortedArr.length; i++) {
        if (nums[i] !== nums[i - 1] + 1) {
            return nums[i] - 1;
        }
    }
}

const missingNumberHashSet = (nums) => {
    const hashSet = new Set(nums);

    for (let i = 0; i <= nums.length; i++) {
        if (!hashSet.has(i)) return i;
    }


    return hashSet;
}

//linear time and complexity

//Gauss Formula
/*We can compute the sum of nums in linear time, and by Gauss' formula, we
can compute the sum of the first n natural numbers in constant time. Therefore,
the number that is missing is simply the result of Gauss' formula minus the sum of nums,
as nums consists of the first n natural numbers minus some number.*/

const missingNumberGauss = (nums) => {
    const expectedSum = nums.length * (nums.length + 1) / 2;
    let actualSum = 0;

    for (let num of nums) {
        actualSum += num;
    }

    return expectedSum - actualSum;
}

const nums = [9,6,4,2,3,5,7,0,1];
console.log(missingNumber(nums));
console.log(nums);

File: HashmapAndSet/missing_number.py
from typing import List


class MissingNumber:
    def missingNumber(self, nums: List[int]) -> int:
        n = len(nums)
        num_set = set(nums)
        for i in range(n + 1):
            if i not in num_set:
                return i
        return None  # This line should never be reached per the problem constraints

    # Time complexity: O(n) where n is the length of the input array.
    # We need to iterate through the input array to create the set,
    # which takes O(n) time.
    # We also need to iterate through the numbers from 0 to n to check for the missing number,
    # which takes O(n) time.
    # Space complexity: O(n) for the set.

    def missingNumberGauss(self, nums: List[int]) -> int:
        n = len(nums)
        expected_sum = n * (n + 1) // 2
        actual_sum = sum(nums)
        return expected_sum - actual_sum

    # Time complexity: O(n) where n is the length of the input array.
    # We need to iterate through the input array to calculate the actual sum,
    # which takes O(n) time.
    # Space complexity: O(1)

    def missingNumberXOR(self, nums: List[int]) -> int:
        n = len(nums)
        missing = n  # Start with n, since we will XOR with indices from 0 to n-1
        for i in range(n):
            missing ^= i ^ nums[i]
        return missing
    
    # Time complexity: O(n) where n is the length of the input array.
    # We need to iterate through the input array once, which takes O(n) time.
    # Space complexity: O(1)
